File: VPSApps/src/lichess.py
""" lichess """
import json
import os
import platform
import urllib
from vpsutils import VpsUtils


def get_token(vps):
    """ Read token from file.
        Args:
            vps: VpsUtils
    """
    file = '_lichesst' if platform.system() == 'Windows' else '.lichesst'
    full_path = os.path.join(vps.home, file)
    with open(full_path, 'r') as lichesst:
        return lichesst.read().rstrip('\n')


def fetch_overview(token):
    """ Fetch overview of ongoing games.
        Args:
            token: lichess token
    """
    try:
        url = 'https://lichess.org/api/account/playing'
        request = urllib.request.Request(
            url, data=None,
            headers={
                'User-Agent': ('Mozilla/5.0 (X11; Ubuntu; Linux x86_64;'
                               'rv:55.0) Gecko/20100101 Firefox/55.0'),
                'Authorization': 'Bearer ' + token
            }
        )
        return urllib.request.urlopen(request).read()
    except urllib.error.URLError:
        return None


def get_my_turn_games(overview):
    """ Get the games in which I am in turn.
        Args:
            overview: the oveview json file
    """
    games = []
    for game in overview["nowPlaying"]:
        if game["isMyTurn"]:
            games.append(game)

    return games


def games_to_html(games):
    """ Create HTML from games.
        Args:
            games: games in which I am in turn
    """
    html = ""
    formatter = "<li><i>{}</i>: {}</li>"
    linkformatter = "<a href=\"https://lichess.org/{}\">Go to game</a>"
    for game in games:
        opponent = game["opponent"]["username"]
        link = linkformatter.format(game["fullId"])
        line = formatter.format(opponent, link)
        html = "{0}{1}".format(html, line)

    return html


def main():
    """ Lichess main method. """
    vps = VpsUtils('Lichess')
    token = get_token(vps)
    jsondata = fetch_overview(token)
    if jsondata is not None:
        overview = json.loads(jsondata.decode('utf-8'))
        games = get_my_turn_games(overview)
        if games:
            html = games_to_html(games)
            vps.html_add('<p>My turn in game vs.:</p><ul>')
            vps.html_add(html)
            vps.html_add('</ul>')
            vps.deliver()


if __name__ == '__main__':
    main()

File: VpsApps/perl/lichess.pl
# Install:
#	apt install libssl-dev
#	File::HomeDir, LWP::Simple, HTTP::Request, MIME::Lite, JSON
#	Bundle::LWP, LWP::Protocol::https
use strict;
use File::HomeDir;
use LWP::Simple;
use LWP::Protocol::https;
use HTTP::Request;
use MIME::Lite;
use JSON;
use POSIX qw(strftime);

sub get_token {
	my $tokenFile = File::HomeDir->my_home . "/.lichesst";
	open my $file, '<', $tokenFile;
	my $token = <$file>;
	close $file;
	$token =~ s/^\s+|\s+$//g;
	return $token;
}

sub get_lichess_json {
	my ($token) = @_;
	my $url = "https://lichess.org/api/account/playing";
	my $ua = LWP::UserAgent->new;
	my $request = HTTP::Request->new(GET => $url);
	$request->header("User-Agent" => "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:55.0) Gecko/20100101 Firefox/55.0");
	$request->header("Authorization" => "Bearer ${token}");
	my $res = $ua->request($request);
	if (!$res->is_success) {
		exit;
	}

	return decode_json($res->decoded_content);
}

sub get_games {
	my ($title, @nowPlaying) = @_;

	my $block = "";
	foreach my $game (@nowPlaying) {
		next if (!$game->{'isMyTurn'});
		my $opponent = $game->{'opponent'};
		my $name = $opponent->{'username'};
		my $gameId = $game->{"gameId"};
		$block = "${block}<li><i>${name}</i>: <a href=\"https://lichess.org/${gameId}\">Go to game</a></li>";
	}

	return $block;
}

sub get_html {
	my ($title, $blocks) = @_;

	return <<"EOF";
<!DOCTYPE html>
<html dir="ltr" lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <meta name="viewport" content="width=device-width" />
    <title>$title</title>
</head>
<body>
<p>My turn in game vs.:</p>
<ul>
$blocks
</ul>
</body>
</html>
EOF
}

sub send_email {
	my ($title, $html) = @_;

	MIME::Lite->send('smtp', 'localhost', Timeout => 60, Port => 25);
	my $msg = MIME::Lite->new(
		"From" => 'bs@leah.schau.dk',
		"To" => 'brian@schau.dk',
		"Subject" => $title,
		"Data" => $html
	);
	$msg->attr("Content-Type" => "text/html");
	$msg->send();
}

my $token = get_token();
my $text = get_lichess_json($token);
my @nowPlaying = @{$text->{'nowPlaying'}};
if (!@nowPlaying) {
	exit;
}

my $title = strftime "Lichess %F %T", localtime;
my $blocks = get_games($title, @nowPlaying);
if (length($blocks) == 0) {
	exit;
}

my $html = get_html($title, $blocks);

send_email($title, $html);
